### ISSUE TITLE
resume training does not work for multi-gpus training

### ISSUE THREAD

**Issue by forever208** [openai/guided-diffusion#23]
I add `--resume_checkpoint $path_to_checkpoint$` to continue the training, it works for a single GPU, but does not work for multi-gpus

the code gets stuck here:

Logging to /proj/ihorse_2021/users/x_manni/guided-diffusion/log9
creating model and diffusion...
creating data loader...
start training...
loading model from checkpoint: /proj/ihorse_2021/users/x_manni/guided-diffusion/log9/model200000.pt...

**Comment by VigneshSrinivasan10** [comment]
@forever208 I have the same problem and the code gets stuck forever. 

On further investigation, I found that the test script `image_sample.py` reloads the model. 
Here is the difference: 
The test script reloads the model before placing the model on cuda. 
However, the training script already has the model on cuda and this leads to the problem of getting stuck there. 
Upon debugging, I found the code to get stuck on this line in `dist_util.py`: 
https://github.com/openai/guided-diffusion/blob/27c20a8fab9cb472df5d6bdd6c8d11c8f430b924/guided_diffusion/dist_util.py#L67 

It is not clear to me why this fails. 
Any pointers in fixing this problem would be greatly appreciated. 
Thanks in advance.

**Comment by bahjat-kawar** [comment]
@VigneshSrinivasan10  Not sure if it's the same issue, but I also found the code to stop at the same line when running `classifier_train.py`.
I fixed the issue by indenting `load_state_dict` out of the `if dist.get_rank() == 0` block on [line 51](https://github.com/openai/guided-diffusion/blob/main/scripts/classifier_train.py#L51).
Hope this helps.

**Comment by VigneshSrinivasan10** [comment]
@bahjat-kawar Thanks for tip and sorry for the delay in my response. 
Your suggestion fixed the problem.

**Comment by VigneshSrinivasan10** [comment]
@bahjat-kawar Although the model reloading was successful, I still face loss values going to NAN after retraining for a few iterations. All the three .pt files were reloaded, but this issue still persists. I assumed the opt.pt file should have some information of the optimizer parameters which should help continue the training. 

Did you also face this issue?

**Comment by JiamingLiu-Jeremy** [comment]
@VigneshSrinivasan10  I met the similar problem. Any progress on fixing this issue ?

**Comment by ONobody** [comment]
@forever208 Hello, do you use opt, model or ema's. pt file when using resume_checkpoint?
Or put their. pt files in a folder as the path of resume_checkpoint.

**Comment by forever208** [comment]
@ON I use model to do resume training (where both ema and opt will be loaded). use ema to do sampling

**Comment by ONobody** [comment]
@forever208 When I continue to train,
such as  python image_train.py --resume_checkpoint  path/modelXX.pt? think you very  much

**Comment by forever208** [comment]
@ONobody exactly
If you have further trouble, take a look at this pull request [Fix resumed model training for Multi-GPUs](https://github.com/openai/guided-diffusion/pull/64)

**Comment by ONobody** [comment]
@forever208 Thank you very much.

**Comment by ONobody** [comment]
@forever208 This is my training parameter.
![1](https://user-images.githubusercontent.com/108787020/223011635-84efe422-6c6e-4f12-8926-f6f95cfdb4d6.png)
I use a single gpu.  But there is still an error message
![2](https://user-images.githubusercontent.com/108787020/223011747-4036c8c6-77f7-4f09-b1b1-f6ce90688e45.png)

**Comment by ONobody** [comment]
@forever208 Hello, I would like to ask how to train Classifier guidance on my own data set.
Do you need to change any codes?
I always make mistakes.

**Comment by forever208** [comment]
@ONobody I have no experience of using the classifier guidance, sorry for not being able to help you in this case

**Comment by ONobody** [comment]
@forever208 What about the calculation of FID IS and other evaluation indicators?
I don't know how to calculate it here

**Comment by forever208** [comment]
@ONobody  the author provides the instructions: https://github.com/openai/guided-diffusion/tree/main/evaluations

**Comment by ONobody** [comment]
@forever208 
The diffusion I trained is based on my own data set.
How to evaluate this thank you.

**Comment by forever208** [comment]
@ONobody if your own dataset only has one class, you can randomly draw 50k samples to form the reference_batch. Then you generated 50k samples using your trained model. Computing the FID by running the script

```
$ python evaluator.py reference_batch.npz 50k_samples.npz
```

If your own dataset has more than 1 class, you'd better use the whole training set as the reference_batch.

remember to convert your data into .npz format

**Comment by ONobody** [comment]
@forever208 
The picture size of my dataset is not 256.
But the size of the picture I generated is 256.
Do I need to change the picture size of my data set to 256?
thank you

**Comment by forever208** [comment]
@ONobody you have to keep them the same size. For example, your training data must be resized to 256 when doing the training. Then, your model generates a 256*256 sample.

**Comment by ONobody** [comment]
@forever208 When I make an assessment,
Convert your own data set into 256*256.
And then convert it to npz format, right?

**Comment by forever208** [comment]
@ONobody  convert the training data into 256x256 --> training the model --> sampling 50k samples (256x256) from the model --> convert both referench batch (256x256) and 50k samples (256x256) into npz file --> compute FID

**Comment by ONobody** [comment]
@forever208 Thank you very much.
When converting my data set to npz format and then calculating fid.
![图片1](https://user-images.githubusercontent.com/108787020/223601707-34b34645-9c75-4501-a4fb-73d055b0ad19.png)
Is it the wrong way for me to convert?

**Comment by open11012** [comment]
> @VigneshSrinivasan10 Not sure if it's the same issue, but I also found the code to stop at the same line when running `classifier_train.py`. I fixed the issue by indenting `load_state_dict` out of the `if dist.get_rank() == 0` block on [line 51](https://github.com/openai/guided-diffusion/blob/main/scripts/classifier_train.py#L51). Hope this helps.

Thanks! I meet the same problem. It works in  my code!
